org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java

```diff
@@ -200,6 +203,12 @@ public int instrumentAll(final InputStream input,
 
 	private int instrumentZip(final InputStream input,
 			final OutputStream output, final String name) throws IOException {
+		if (generators.getLast() instanceof
+				OfflineInstrumentationCompanionAccessGenerator) {
+			generators
+					.addLast(new
+							OfflineInstrumentationCompanionAccessGenerator());
```

Comment:
instrumentAll was previously thread-safe, but it no longer is due to this addLast when used with OfflineInstrumentationCompanionAccessGenerator, which is also not thread-safe.  This should probably be documented somehow for both classes.